aiida/restapi/resources.py

```diff
@@ -207,6 +208,130 @@ def get(self, id=None, page=None):  # pylint: disable=redefined-builtin,invalid-
 
 
+class QueryBuilder(BaseResource):
+    """
+    Representation of a QueryBuilder REST API resource (instantiated with a queryhelp JSON).
+
+    It supports POST requests taking in JSON :py:func:`~aiida.orm.querybuilder.QueryBuilder.queryhelp`
+    objects and returning the :py:class:`~aiida.orm.querybuilder.QueryBuilder` result accordingly.
+    """
+    from aiida.restapi.translator.nodes.node import NodeTranslator
+
+    _translator_class = NodeTranslator
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+        # HTTP Request method decorators
+        if 'get_decorators' in kwargs and isinstance(kwargs['get_decorators'], (tuple, list, set)):
+            self.method_decorators.update({'post': list(kwargs['get_decorators'])})
+
+    def get(self):  # pylint: disable=arguments-differ
+        """Static return to state information about this endpoint."""
+        data = {
+            'message': (
+                'Method Not Allowed. Use HTTP POST requests to use the AiiDA QueryBuilder. '
+                'POST JSON data, which MUST be a valid QueryBuilder.queryhelp dictionary as a JSON object. '
+                'See the documentation at https://aiida.readthedocs.io/projects/aiida-core/en/latest/topics/'
+                'database.html?highlight=QueryBuilder#the-queryhelp for more information.'
+            ),
+        }
+
+        headers = self.utils.build_headers(url=request.url, total_count=1)
+        return self.utils.build_response(
+            status=405,  # Method Not Allowed
+            headers=headers,
+            data={
+                'method': request.method,
+                'url': unquote(request.url),
+                'url_root': unquote(request.url_root),
+                'path': unquote(request.path),
+                'query_string': request.query_string.decode('utf-8'),
+                'resource_type': self.__class__.__name__,
+                'data': data,
+            },
+        )
+
+    def post(self):
+        """
+        POST method to pass query help JSON.
+
+        This assumes the posted data is a complete and valid QueryBuilder queryhelp JSON object.
+        We could create a RequestParser that completely redefines all valid keys in the queryhelp,
+        but for now, it's much easier to simply pass-through the posted queryhelp and make sure
+        it can be handled by the built-in NodeTranslator.
+
+        This uses the NodeTranslator in order to best return Nodes according to the general AiiDA
+        REST API data format, while still allowing the return of other AiiDA entities.
+
+        :return: QueryBuilder result of AiiDA entities in "standard" REST API format.
+        """
+        # pylint: disable=protected-access
+        self.trans._query_help = json.loads(request.json) if isinstance(request.json, str) else request.json
+        assert isinstance(self.trans._query_help, dict), (
+            'POSTed data MUST be a valid QueryBuilder.queryhelp dictionary. '
```

Comment:
I discussed this earlier and decided against it for now due to the amount of extra work involved versus what is gained.

I don't have experience with the `voluptuous` package, but I was thinking of using `pydantic` or a simple data class.
I'd rather this kind of schema comes directly from a `QueryBuilder` class, since otherwise I will be defining a `QueryBuilder` queryhelp schema outside the class, which is a slippery slope towards difficult code maintenance.

Then, if it should be included in the `QueryBuilder` class it should be a complete schema, hence the amount of work suddenly increased drastically, while the result for a user will be the same - an error. The message might be clearer in pinpointing how the queryhelp JSON is wrong, but at this stage I don't think the work is worth it. Especially since I expect the usage here to come from either a generated queryhelp JSON from a `QueryBuilder` object or auto-generated through a parser, i.e., no manual user input. However, this is of course my assumption and might be totally wrong.

Reply:
Ok to not include it in this PR, but please open an issue for this, mentioning this line.

It will indeed lead to better error messages. Even if people are getting their queryhelp JSONs from AiiDA, this will be useful as soon as anything in the queryhelp format changes.